jaxdecomp/_src/fft.py

```diff
@@ -36,290 +54,423 @@ def _str_to_fft_type(s: str) -> xla_client.FftType:
-  elif isinstance(fft_type, xla_client.FftType):
-    typ = fft_type
-  else:
-    raise TypeError(f"Unknown FFT type value '{fft_type}'")
-
-  if typ in [xla_client.FftType.RFFT, xla_client.FftType.IRFFT]:
-    raise TypeError("only complex FFTs are currently supported through pfft.")
+class FFTPrimitive(BasePrimitive):
+  """
+  Custom primitive for FFT operations.
+  """
 
-  (x,) = promote_dtypes_complex(x)
+  name = "fft"
+  multiple_results = False
+  impl_static_args = (1, 2)
+  inner_primitive = None
+  outer_primitive = None
 
-  return sfft_p.bind(
-      x, fft_type=typ, pdims=pdims, global_shape=global_shape, adjoint=adjoint)
-
-
-def sfft_abstract_eval(x, fft_type, pdims, global_shape, adjoint):
-
-  # TODO(Wassim) : this only handles cube shapes
-  # This function is called twice once with the global array and once with the local slice shape
-
-  # Figure out what is the pencil decomposition at the output
-  axis = 0
-  if fft_type in [xla_client.FftType.RFFT, xla_client.FftType.FFT]:
-    axis = 2
-
-  output_shape = None
-  match fft_type:
-    case xla_client.FftType.FFT:
-      # FFT is X to Y to Z so Z-Pencil is returned
-      # Except if we are doing a YZ slab in which case we return a Y-Pencil
-      transpose_shape = (1, 2, 0)
-      transposed_pdims = pdims
-    case xla_client.FftType.IFFT:
-      # IFFT is Z to X to Y so X-Pencil is returned
-      # In YZ slab case we only need one transposition back to get the X-Pencil
-      transpose_shape = (2, 0, 1)
-      transposed_pdims = pdims
-    case _:
-      raise TypeError("only complex FFTs are currently supported through pfft.")
+  @staticmethod
+  def abstract(x: Array, fft_type: xla_client.FftType, pdims: Tuple[int, int],
+               global_shape: Tuple[int, int,
+                                   int], adjoint: bool) -> ShapedArray:
+    """
+    Abstract function to compute the shape of FFT output.
+
+    Parameters
+    ----------
+    x : Array
+      Input array.
+    fft_type : xla_client.FftType
+      Type of FFT operation.
+    pdims : Tuple[int, int]
+      Parallel dimensions.
+    global_shape : Tuple[int, int, int]
+      Global shape of the array.
+    adjoint : bool
+      Whether to compute the adjoint FFT.
+
+    Returns
+    -------
+    ShapedArray
+      Shape of the output array.
+    """
+    if global_shape == x.shape:
+      return FFTPrimitive.outer_abstract(x, fft_type=fft_type, adjoint=adjoint)
+
+    match fft_type:
+      case xla_client.FftType.FFT:
+        # FFT is X to Y to Z so Z-Pencil is returned
+        # Except if we are doing a YZ slab in which case we return a Y-Pencil
+        transpose_shape = (1, 2, 0)
+        transposed_pdims = pdims
+      case xla_client.FftType.IFFT:
+        # IFFT is Z to X to Y so X-Pencil is returned
+        # In YZ slab case we only need one transposition back to get the X-Pencil
+        transpose_shape = (2, 0, 1)
+        transposed_pdims = pdims
+      case _:
+        raise TypeError(
```

Comment:
Make sure to add Python ≥ 3.10 in the requirements to be allowed to use such pattern.

Reply:
good point